torch/csrc/jit/passes/memory_planning.h

```diff
@@ -0,0 +1,151 @@
+  return lhs.offset == rhs.offset && lhs.size == rhs.size;
+}
+
+struct regionSizeCmp {
+  bool operator()(const MemRegion& reg1, const MemRegion& reg2) const {
+    return reg1.size == reg2.size ? reg1.offset < reg2.offset
+                                  : reg1.size < reg2.size;
+  }
+};
+
+struct regionOffsetCmp {
+  bool operator()(const MemRegion& reg1, const MemRegion& reg2) const {
+    return reg1.offset == reg2.offset ? reg1.size < reg2.size
+                                      : reg1.offset < reg2.offset;
+  }
+};
+
+bool overlapMemRegion(const MemRegion& reg1, const MemRegion& reg2);
+
+struct UniqueLiveRange {
+  LiveRange lvr;
+  std::string id;
+};
+
+bool overlapLiveRange(
+    const UniqueLiveRange& ulvr1,
+    const UniqueLiveRange& ulvr2);
+
+inline std::ostream& operator<<(std::ostream& str, UniqueLiveRange rhs) {
+  return str << "{id: " << rhs.id << ", lvr: " << rhs.lvr << "}";
+}
+
+inline bool operator==(const UniqueLiveRange lhs, const UniqueLiveRange rhs) {
+  return lhs.lvr == rhs.lvr && lhs.id == rhs.id;
+}
+
+struct liveRangeStartCmp {
+  bool operator()(const UniqueLiveRange& u1, const UniqueLiveRange& u2) const {
+    return u1.lvr.begin == u2.lvr.begin
+        ? (u1.lvr.end == u2.lvr.end ? u1.id < u2.id : u1.lvr.end < u2.lvr.end)
+        : u1.lvr.begin < u2.lvr.begin;
+  }
+};
+
+struct liveRangeEndCmp {
+  bool operator()(const UniqueLiveRange& u1, const UniqueLiveRange& u2) const {
+    return u1.lvr.end == u2.lvr.end
+        ? (u1.lvr.begin == u2.lvr.begin ? u1.id < u2.id
+                                        : u1.lvr.begin < u2.lvr.begin)
+        : u1.lvr.end < u2.lvr.end;
+  }
+};
+
+template <typename Value>
+using SortedLiveRangeMap = std::map<UniqueLiveRange, Value, liveRangeStartCmp>;
+struct TORCH_API MemAllocation {
+  UniqueLiveRange ulvr;
+  MemRegion reg;
+};
+
+inline std::ostream& operator<<(std::ostream& str, MemAllocation rhs) {
+  return str << rhs.ulvr << ", " << rhs.reg;
+}
+
+inline bool operator==(const MemAllocation lhs, const MemAllocation rhs) {
+  return lhs.ulvr == rhs.ulvr && lhs.reg == rhs.reg;
+}
+
+inline bool valid_add(size_t a, size_t b) {
+#if defined(_MSC_VER)
+  return a + b >= a;
+#else
+  size_t _carry = 0;
+  return !__builtin_add_overflow(a, b, &_carry);
+#endif
+}
+
+inline bool valid_sub(size_t a, size_t b) {
+#if defined(_MSC_VER)
+  return a >= b;
+#else
+  size_t _carry = 0;
+  return !__builtin_sub_overflow(a, b, &_carry);
+#endif
+}
+
+c10::optional<size_t> computeStorageSize(const Value& value);
+
+TORCH_API bool hasOutVariant(Node* node);
+
+TORCH_API std::
+    pair<size_t, FastMap<const Value*, std::pair<UniqueLiveRange, size_t>>>
+    planMemory(const std::shared_ptr<Graph>&, Strategy);
+
+} // namespace jit
+} // namespace torch
+
+namespace std {
+
+using namespace torch::jit;
+
+template <>
+struct hash<MemRegion> {
+  size_t operator()(const MemRegion& reg) const {
+    return std::hash<size_t>()(reg.offset) ^
```

Comment:
should we just use `c10::get_hash(reg.offset, reg.size)` ? (have not thought that hard abt the hash function here so tell me if im wrong). also if we suspect only the first 32 bits of each value will get used we could just combine them together [first 32 bits reg.offset, first 32 bits reg.size]

Reply:
not sure how `get_hash` works (will take a look) but the `<<1` is actually necessary since other wise for `offset == size` you would get a hash of 0 (`k ^ k = 0`). don't quite understand the second part - why do we think only the first 32 bits will be used?

Reply:
`<< 1` won't completely solve the problem you are referring to, will it? I would also prefer to use `c10::get_hash` instead of inventing your own hash function unless data suggest this is performance critical.